aws/resource_aws_dx_gateway_association_migrate.go

```diff
@@ -0,0 +1,101 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/directconnect"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+)
+
+func resourceAwsDxGatewayAssociationResourceV0() *schema.Resource {
+	return &schema.Resource{
+		Schema: map[string]*schema.Schema{
+			"allowed_prefixes": {
+				Type:     schema.TypeSet,
+				Optional: true,
+				Computed: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+
+			"associated_gateway_id": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				Computed:      true,
+				ForceNew:      true,
+				ConflictsWith: []string{"associated_gateway_owner_account_id", "proposal_id", "vpn_gateway_id"},
+			},
+
+			"associated_gateway_owner_account_id": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				Computed:      true,
+				ForceNew:      true,
+				ValidateFunc:  validateAwsAccountId,
+				ConflictsWith: []string{"associated_gateway_id", "vpn_gateway_id"},
+			},
+
+			"associated_gateway_type": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"dx_gateway_association_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"dx_gateway_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"dx_gateway_owner_account_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"proposal_id": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				ForceNew:      true,
+				ConflictsWith: []string{"associated_gateway_id", "vpn_gateway_id"},
+			},
+
+			"vpn_gateway_id": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				ForceNew:      true,
+				ConflictsWith: []string{"associated_gateway_id", "associated_gateway_owner_account_id", "proposal_id"},
+				Deprecated:    "use 'associated_gateway_id' argument instead",
+			},
+		},
+	}
+}
+
+func resourceAwsDxGatewayAssociationStateUpgradeV0(rawState map[string]interface{}, meta interface{}) (map[string]interface{}, error) {
+	conn := meta.(*AWSClient).dxconn
+
+	log.Println("[INFO] Found Direct Connect gateway association state v0; migrating to v1")
+
+	// dx_gateway_association_id was introduced in v2.8.0. Handle the case where it's not yet present.
+	if _, ok := rawState["dx_gateway_association_id"]; !ok {
```

Comment:
I'm not sure why, but starting from a version 3 state (Terraform 0.11.14), it appears this attribute is coming across in `rawState` as a present key in the map with `nil` interface value: `"dx_gateway_association_id":interface {}(nil)`, which prevented the lookup logic from occurring for me locally.

Changing this to the below allowed it to succeed in both Terraform 0.11.14 and Terraform 0.12.12 with this branch:

```suggestion
	if v, ok := rawState["dx_gateway_association_id"]; !ok || v == nil {
```